### PR TITLE
Authentication and Credentials dialog

### PR DESCRIPTION
**Pull request by stefwalter**
This implements the authentication and credentials dialog described here: https://github.com/cockpit-project/cockpit/wiki/Feature:-Key-based-authentication

### CODE REVIEW

**Comment by stefwalter** [comment]
Needs tests and documentation update.

**Comment by andreasn** [comment]
No on/off switch for "Use my password for privileged tasks and to connect to other machines"?

How do I I access the add button?

**Comment by stefwalter** [comment]
> How do I I access the add button?

That functionality is not yet present in cockpit-ws.

**Comment by stefwalter** [comment]
Depends on #2882

**Comment by stefwalter** [comment]
Will do key generation and proper deauthorize UI in other pull requests.

**Comment by stefwalter** [comment]
Depends on #2882  and #2894

**Comment by stefwalter** [comment]
Rebased onto #2894

File: pkg/shell/index.html [hunk 1]
```diff
@@ -162,6 +164,148 @@ <h4 class="modal-title" id="error-popup-title"></h4>
...
+                                            <input type="radio" autocomplete="off" checked>
+                                            <span translatable="yes">Off</span>
+                                        </label>
+                                    </div>
+                                </td>
+                            </tr>
+                            <tr class="listing-head credential-panel">
+                                <td colspan="3">
+                                    <ul class="nav nav-tabs nav-tabs-pf">
+                                        <li class="active"><a href="#">Details</a></li>
```
Comment (petervo) [inline pkg/shell/index.html]: Translatable

File: pkg/shell/index.html [hunk 2]
```diff
@@ -162,6 +164,148 @@ <h4 class="modal-title" id="error-popup-title"></h4>
...
+                                        <li><a href="#">Public Key</a></li>
```
Comment (petervo) [inline pkg/shell/index.html]: Translatable

File: pkg/shell/index.html [hunk 3]
```diff
@@ -162,6 +164,148 @@ <h4 class="modal-title" id="error-popup-title"></h4>
...
+                                        <li><a href="#">Password</a></li>
```
Comment (petervo) [inline pkg/shell/index.html]: Translatable

File: pkg/shell/credentials.js [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+            var two_pass = body.find(".credential-two").val();
+            if (old_pass === undefined || new_pass === undefined || two_pass === undefined)
+                throw "invalid password fields";
+
+            keys.change(key.name, old_pass, new_pass, two_pass)
+                .always(function(ex) {
+                    body.find("input button").prop("disabled", false);
+                })
+                .done(function() {
+                    body.find("li a").first().click();
```
Comment (petervo) [inline pkg/shell/credentials.js]: We should clear the password fields on success

File: pkg/shell/credentials.js [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+
+            var value = $(this).onoff("value");
+
+            /* Key needs to be loaded, show load UI */
+            if (value && !key.loaded) {
+                body.addClass("open").addClass("unlock");
+
+            /* Key needs to be unloaded, do that directly */
+            } else if (!value && key.loaded) {
+                keys.unload(key.name)
```
Comment (petervo) [inline pkg/shell/credentials.js]: Should we collapse the panel and clear any fields after unloading.

File: pkg/shell/index.html [hunk 6]
```diff
@@ -162,6 +164,148 @@ <h4 class="modal-title" id="error-popup-title"></h4>
...
+                                <td translatable="yes" colspan="2">Use my password for privileged tasks and to connect to other machines</td>
+                                <td class="listing-actions">
+                                    <button class="btn btn-default pficon pficon-close" id="deauthorize-item">
+                                    </button>
+                                </td>
+                            </tr>
+                            <tr>
+                                <td translatable="yes" colspan="2">Use the following keys to authenticate against other systems</td>
+                                <td class="listing-actions">
+                                    <button class="btn btn-primary fa fa-plus disabled">
```
Comment (petervo) [inline pkg/shell/index.html]: I'm guessing generating a new key is a different pull request. Should we remove this button until then?

File: pkg/shell/credentials.js [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+            var key = keys.items[id];
+            if (!key || !key.name)
+                return;
+
+            body.find("input button").prop("disabled", true);
+
+            var password = body.find(".credential-password").val();
+            keys.load(key.name, password)
+                .always(function(ex) {
+                    body.find("input button").prop("disabled", false);
```
Comment (petervo) [inline pkg/shell/credentials.js]: we should probably clear the password field here as well.

File: pkg/shell/index.html (hunk 1 above)
Comment (stefwalter) [inline pkg/shell/index.html]: Fixed.

File: pkg/shell/index.html (hunk 2 above)
Comment (stefwalter) [inline pkg/shell/index.html]: Fixed.

File: pkg/shell/index.html (hunk 3 above)
Comment (stefwalter) [inline pkg/shell/index.html]: Fixed.

File: pkg/shell/credentials.js (hunk 4 above)
Comment (stefwalter) [inline pkg/shell/credentials.js]: Done.

File: pkg/shell/credentials.js (hunk 5 above)
Comment (stefwalter) [inline pkg/shell/credentials.js]: Done.

File: pkg/shell/index.html (hunk 6 above)
Comment (stefwalter) [inline pkg/shell/index.html]: Commented out.

File: pkg/shell/credentials.js (hunk 7 above)
Comment (stefwalter) [inline pkg/shell/credentials.js]: Done.

File: test/check-keys [hunk 8]
```diff
@@ -150,4 +150,98 @@ class TestKeys(MachineCase):
...
+        m.execute("chmod 600 /home/admin/.ssh/*")
+        m.execute("chown -R admin:admin /home/admin/.ssh")
+
+        self.login_and_go()
+
+        id_rsa = "2048 SHA256:SRvBhCmkCEVnJ6ascVH0AoVEbS3nPbowZkNevJnXtgw"
+        id_dsa = "1024 SHA256:x6S6fxMuEyqhpwNRAIK7ms6bZDY6xK9wzdDr2kCaWVY"
+
+        keys = list_keys()
+        self.assertIn(id_rsa, keys)
```
Comment (petervo) [inline test/check-keys]: This won't pass on atomic with local-ssh. Should we adjust the test? or skip for atomic.

**Comment by stefwalter** [comment]
TestShutdownRestart is a flakey test #2915

File: test/check-keys (hunk 8 above)
Comment (stefwalter) [inline test/check-keys]: Skipping

**Comment by stefwalter** [comment]
Rebased

**Comment by stefwalter** [comment]
Added unittest import to check-keys.

**Pull request closed without merging**